### PR TITLE
Improve chats and communities unmute logic

### PR DESCRIPTION
**Pull request by Parveshdhull**
### Summary

Improvements/Bug fixes
- Execute the initial check immediately upon starting to unmute chats for which the mute timer has expired.
- If a user mutes a chat for 1 minute at 11:12:33, the client displays the mute status as lasting until 11:13. Consequently, the chat should be unmuted as soon as 11:14. Currently, the routine does not run precisely at the top of the minute (hh:mm:00), leading to inconsistencies where the mute timer may expire, but the chat remains muted. This PR ensures that the unmute timer is synchronized to run exactly on the 00 seconds mark.

Required for https://github.com/status-im/status-mobile/issues/20977
Status mobile PR: https://github.com/status-im/status-mobile/pull/21052

Closes #

### CODE REVIEW

File: protocol/messenger.go [hunk 1]
```diff
@@ -1584,56 +1584,70 @@ func (m *Messenger) watchConnectionChange() {
...
+			// Execute the check immediately upon starting
+			response := &MessengerResponse{}
+			currTime := time.Now()
+
+			m.allChats.Range(func(chatID string, c *Chat) bool {
+				chatMuteTill := c.MuteTill
+				if currTime.After(chatMuteTill) && !chatMuteTill.Equal(time.Time{}) && c.Muted {
+					err := m.persistence.UnmuteChat(c.ID)
+					if err != nil {
+						m.logger.Info("Error", zap.Any("Couldn't unmute chat", err))
```
Comment (qfrank) [inline protocol/messenger.go]: will `m.logger.Warn("watchChatsToUnmute error", zap.Any...)` be better? 🤔

File: protocol/messenger.go [hunk 2]
```diff
@@ -1584,56 +1584,70 @@ func (m *Messenger) watchConnectionChange() {
...
+				return true
+			})
+
+			if !response.IsEmpty() {
+				signal.SendNewMessages(response)
+			}
+
+			// Calculate the time until the next whole minute
+			now := time.Now()
+			waitDuration := time.Until(now.Truncate(time.Minute).Add(time.Minute))
```
Comment (qfrank) [inline protocol/messenger.go:1610]: maybe move this line to `1595` and caculate the duration based on `currTime` would be more accurate?

File: protocol/messenger.go [hunk 3]
```diff
@@ -1584,56 +1584,70 @@ func (m *Messenger) watchConnectionChange() {
...
-			select {
-			case <-time.After(1 * time.Minute):
-				response, err := m.CheckCommunitiesToUnmute()
-				if err != nil {
-					return
-				}
+			// Execute the check immediately upon starting
+			response, err := m.CheckCommunitiesToUnmute()
+			if err != nil {
+				m.logger.Info("Error", zap.Any("Couldn't unmute communities", err))
```
Comment (qfrank) [inline protocol/messenger.go]: same as above

File: protocol/messenger.go [hunk 4]
```diff
@@ -1584,56 +1584,70 @@ func (m *Messenger) watchConnectionChange() {
...
+			} else if !response.IsEmpty() {
+				signal.SendNewMessages(response)
+			}
 
-				if !response.IsEmpty() {
-					signal.SendNewMessages(response)
-				}
+			// Calculate the time until the next whole minute
+			now := time.Now()
+			waitDuration := time.Until(now.Truncate(time.Minute).Add(time.Minute))
```
Comment (qfrank) [inline protocol/messenger.go:1638]: `waitDuration` seems will add the execution delay with `m.CheckCommunitiesToUnmute()` , would it be more accurate by moving this line before 1636 ?

**Review by qfrank — APPROVED** [review]

File: protocol/messenger.go (hunk 4 above)
Comment (Parveshdhull) [reply protocol/messenger.go:1638]: Thank you @qfrank for reviewing the PR.

The difference is in milliseconds, but having a wait duration in the end is more accurate.
The calculation of wait duration checks how much time is left between the next exact minute. Calculating this just before calling `time.After` will avoid any delay added by other stuff.

image

File: protocol/messenger.go (hunk 1 above)
Comment (Parveshdhull) [reply protocol/messenger.go]: thank you, done

File: protocol/messenger.go (hunk 4 above)
Comment (qfrank) [reply protocol/messenger.go:1638]: Oh, I ignored `Truncate`, pls ignore my previous view regarding this!

**Review by ilmotta — APPROVED** [review]
LGTM @Parveshdhull. There are some tests covering the unmute logic here https://github.com/status-im/status-go/blob/c257874be93889cd4fbfd511fea43dc976eb01d6/protocol/messenger_mute_test.go#L74

If possible, consider adding a new test covering the time boundaries we are using now if it's not too complicated. This SO shares practical solutions to how to control the time in the tests https://stackoverflow.com/a/58145008
